File: Problems/0100_0199/0116_Populating_Next_Right_Pointers_in_Each_Node/Project_Java/Operate_Node.java
import java.util.*;

public class Operate_Node {
    public Node set_Node(String[] flds) {
         return set_Node(flds, 0, 0);
    }

    public Node set_Node(String[] flds, int depth, int pos) {
        if (flds.length == 0)
            return null;

        int cur_pos = 0;
        for (int i = 0; i < depth; ++i)
            cur_pos += (int)Math.pow(2, i);
        
        if (cur_pos + pos > flds.length - 1)
            return null;
        
        if (flds[cur_pos + pos].equals("null"))
            return null;

        try {
            Node node = new Node(Integer.parseInt(flds[cur_pos + pos]));
            node.left = set_Node(flds, depth + 1, 2*pos);
            node.right = set_Node(flds, depth + 1, 2*pos + 1);

            return node;
        } catch (Exception e) {
            System.out.println("\n" +  e + "\n" +
                              "set_Node() Error ... flds[" + Integer.toString(cur_pos + pos) + "] = " + flds[cur_pos + pos] + "\n");
            System.exit(1);

            return null;
        }
    }

    List<String> resultStr;

    public String output_Node(Node node) {
        resultStr = new ArrayList<String>();

        String outStr = set_ResultStr(node, 0);
        resultStr.clear();

        return outStr;
    }

    private String set_ResultStr(Node node, int n) {
        if (node == null)
            return "";
        
        if (resultStr.size() <= n)
            resultStr.add("(" + Integer.toString(node.val) + ")");
        else
            resultStr.set(n, resultStr.get(n) + ",(" + Integer.toString(node.val) + ")");

        if (node.left != null)
            set_ResultStr(node.left, n + 1);
        if (node.right != null)
            set_ResultStr(node.right, n + 1);

        String outputStr = "";
        for(String s : resultStr) {
            outputStr += s + "\n";
        }

        return outputStr;
    }

    public String output_Node_with_next(Node node) {
        resultStr = new ArrayList<String>();

        String outStr = set_ResultStr_with_next(node, 0);
        resultStr.clear();

        return outStr;
    }

    private String set_ResultStr_with_next(Node node, int n) {
        if (node == null)
            return "";
        
        if (resultStr.size() <= n)
            resultStr.add("(" + Integer.toString(node.val) + ")");
        else
            resultStr.set(n, resultStr.get(n) + ",(" + Integer.toString(node.val) + ")");

        if (node.next == null)
                resultStr.set(n, resultStr.get(n) + ",(#)");

        if (node.left != null)
            set_ResultStr_with_next(node.left, n + 1);
        if (node.right != null)
            set_ResultStr_with_next(node.right, n + 1);

        String outputStr = "";
        for(String s : resultStr) {
            outputStr += s + "\n";
        }

        return outputStr;
    }


    public String Tree2str(Node t) {
        if (t == null)
            return "";

        String resultStr = Integer.toString(t.val);

        if (t.left == null && t.right == null)
            return resultStr;

        resultStr += "(" + Tree2str(t.left) + ")";
        if (t.right != null)
            resultStr += "(" + Tree2str(t.right) + ")";

        return resultStr;
    }
}

File: Problems/0100_0199/0116_Populating_Next_Right_Pointers_in_Each_Node/Project_Go/solution/Operate_Node.go
package solution

import (
	"fmt"
	"math"
	"strconv"
)

func setNode(flds []string) *Node {
	return setNode2(flds, 0, 0)
}

func setNode2(flds []string, depth int, pos int) *Node {
	if len(flds) == 0 {
		return nil
	}

	curPos := 0
	for i := 0; i < depth; i++ {
		curPos += int(math.Pow(2.0, float64(i)))
	}

	if curPos+pos > len(flds)-1 {
		return nil
	}

	if flds[curPos+pos] == "null" {
		return nil
	}

	node := new(Node)
	val, err := strconv.Atoi(flds[curPos+pos])

	if err != nil {
		fmt.Printf("setNode() Error ... flds[%d] = %s", curPos+pos, flds[curPos+pos])
		return nil
	}

	node.Val = val
	node.Left = setNode2(flds, depth+1, 2*pos)
	node.Right = setNode2(flds, depth+1, 2*pos+1)

	return node
}

var resultStr = []string{}

func outputNode(node *Node) string {
	resultStr = []string{}

	outStr := setResultStr(node, 0)
	resultStr = nil

	return outStr
}

func setResultStr(node *Node, n int) string {
	if node == nil {
		return ""
	}

	if len(resultStr) <= n {
		resultStr = append(resultStr, "("+strconv.Itoa(node.Val)+")")
	} else {
		resultStr[n] = resultStr[n] + ",(" + strconv.Itoa(node.Val) + ")"
	}

	if node.Left != nil {
		setResultStr(node.Left, n+1)
	}
	if node.Right != nil {
		setResultStr(node.Right, n+1)
	}

	outputStr := ""
	for _, s := range resultStr {
		outputStr += s + "\n"
	}

	return outputStr
}

func outputNode_with_next(node *Node) string {
	resultStr = []string{}

	outStr := setResultStr_with_next(node, 0)
	resultStr = nil

	return outStr
}

func setResultStr_with_next(node *Node, n int) string {
	if node == nil {
		return ""
	}

	if len(resultStr) <= n {
		resultStr = append(resultStr, "("+strconv.Itoa(node.Val)+")")
	} else {
		resultStr[n] = resultStr[n] + ",(" + strconv.Itoa(node.Val) + ")"
	}

	if node.Next == nil {
		resultStr[n] = resultStr[n] + ",(#)"
	}

	if node.Left != nil {
		setResultStr_with_next(node.Left, n+1)
	}
	if node.Right != nil {
		setResultStr_with_next(node.Right, n+1)
	}

	outputStr := ""
	for _, s := range resultStr {
		outputStr += s + "\n"
	}

	return outputStr
}

func Tree2str(t *Node) string {
	if t == nil {
		return ""
	}

	resultStr := strconv.Itoa(t.Val)

	if t.Left == nil && t.Right == nil {
		return resultStr
	}

	resultStr += "(" + Tree2str(t.Left) + ")"
	if t.Right != nil {
		resultStr += "(" + Tree2str(t.Right) + ")"
	}

	return resultStr
}
